pkg/kubelet/cm/deviceplugin/endpoint.go

```diff
@@ -36,7 +37,7 @@ type endpoint interface {
 	run()
 	stop()
 	allocate(devs []string) (*pluginapi.AllocateResponse, error)
-	getDevices() []pluginapi.Device
+	getDevices() map[string]pluginapi.Device
```

Comment:
Any reason for making this a map ?

Reply:
It took me quite some time to figure out that this needs a change.

In the previous implementation, getDevices() return a list/slice. If you search the code for invocations of this method, you can see that only the Manager interface wants a list. In all other cases, a map is needed.
There was even a case where we duplicate this logic with another mutex lock:

```
  func (e) run() {
       devices := make(map ...)
       e.mutex.lock()
       for _, d := range.devices {
          devices[d.id] = d
       }
      e.mutex.unlock()
```

After making this change, the only tax added is the Manager.Devices() call, which has to translate the map to a slice. All other cases when we need the device map, we just call getDevices().

So ... it saves the unnecessary back-forth conversion, it saves unnecessary locks and simpifies the code.

Reply:
It makes sense, I've actually unconsciously done the same thing in #52616 but masked it behind an interface.